### PR TITLE
Fix user roles

### PR DESCRIPTION
**Pull request by darrenoh**

### CODE REVIEW

File: tests/data/menu_example/menu_example.module [hunk 1]
```diff
@@ -377,7 +377,7 @@ function menu_example_permission() {
  *   True if the acting user has the role specified.
  */
 function menu_example_custom_access($role_name) {
-  $access_granted = in_array($role_name, $GLOBALS['user']->roles);
+  $access_granted = array_key_exists($role_name, $GLOBALS['user']->roles);
   return $access_granted;
```
Comment (mglaman) [inline tests/data/menu_example/menu_example.module:381]: If this didn't work, we need to fix the `GlobalUser` and how it returns `roles` as this was the previous code from Drupal 7.

File: src/User/GlobalUser.php [hunk 2]
```diff
@@ -85,7 +86,7 @@ public function getLastAccessedTime()
     public function __get(string $name)
     {
         return match ($name) {
-            'roles' => $this->getRoles(),
+            'roles' => array_combine($this->getRoles(), array_map([Role::class, 'load'], $this->getRoles())),
             'name' => $this->getAccountName(),
             default => null,
```
Comment (mglaman) [inline src/User/GlobalUser.php:91]: In D7 `roles` was an array of role names, correct? So we shouldn't be loading the roles.

File: src/Entity/EntityTypeManager.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * @return EntityTypeInterface[]
+     */
+    protected function findDefinitions(): array
+    {
+        $definitions = $this->getDiscovery()->getDefinitions();
+        if (isset($definitions['user_role']) && $definitions['user_role'] instanceof EntityTypeInterface) {
+            $definitions['user_role']->setClass(Role::class);
+        }
```
Comment (mglaman) [inline src/Entity/EntityTypeManager.php:19]: We're overriding this method to allow altering the definition, correct? Which is preventing us from calling `parent::findDefinitions()` and copying all of its code.

We should override `alterDefinitions` and perform this logic there.

File: src/Entity/Role.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        };
+    }
+
+    public function __isset(string $name): bool
+    {
+        return match ($name) {
+            'rid', 'name' => true,
+            default => false
+        };
+    }
```
Comment (mglaman) [inline src/Entity/Role.php:38]: does this need  match, can't we call `isset($this->$name)`

File: src/functions/user.php [hunk 5]
```diff
@@ -21,3 +23,8 @@ function user_is_logged_in(): bool
 {
     return \Drupal::currentUser()->isAuthenticated();
 }
+
+function user_role_load(string $rid): ?RoleInterface
+{
+    return Role::load($rid);
```
Comment (mglaman) [inline src/functions/user.php:29]: Can we use long-form `\Drupal::entityTypeManager()->getStorage('user_role')->load($rid);`

File: src/User/GlobalUser.php (hunk 2 above)
Comment (darrenoh) [reply src/User/GlobalUser.php:91]: I added a stringable role class, so treating roles as strings returns the role names.

Comment (mglaman) [reply src/User/GlobalUser.php:91]: But, it breaks existing `in_array` checks, which might be common in D7 code, to see if the user has a role (versus having a permission)

File: src/Entity/Role.php (hunk 4 above)
Comment (darrenoh) [reply src/Entity/Role.php:38]: D7 roles have only rid and name properties.

Comment (mglaman) [reply src/Entity/Role.php:38]: Ah, good to know!

File: src/User/GlobalUser.php (hunk 2 above)
Comment (darrenoh) [reply src/User/GlobalUser.php:91]: Didn't test that. I assumed this function wasn't used for that.

File: src/functions/user.php (hunk 5 above)
Comment (darrenoh) [reply src/functions/user.php:29]: I suppose so. I just copied code I found in core.

File: tests/data/menu_example/menu_example.module (hunk 1 above)
Comment (darrenoh) [reply tests/data/menu_example/menu_example.module:381]: In D7, roles is keyed by numeric role ID. In D10, roles is keyed by role name and the values are role objects.

File: src/functions/user.php (hunk 5 above)
Comment (darrenoh) [reply src/functions/user.php:29]: For clarity I changed this to use our extended Role class.

**Review by mglaman — APPROVED** [review]

**Pull request merged**